src/CodeGen_X86.cpp

```diff
@@ -306,10 +306,249 @@ void CodeGen_X86::visit(const Select *op) {
 
 }
 
+bool CodeGen_X86::try_visit_float16_cast(const Cast* op) {
+    Type destTy = op->type;
+    Type srcTy = op->value.type();
+
+    // half -> single/double
+    if (destTy.is_float() && srcTy.is_float() && srcTy.bits == 16 && destTy.bits > 16) {
+        internal_assert(destTy.bits == 64 || destTy.bits == 32) << "Unexpected float type\n";
+
+        if (target_needs_software_cast_from_float16_to(destTy)) {
+            // Let parent class codegen calling software implementation of cast
+            return false;
+        }
+
+        // Use @llvm.x86.vcvtph2ps.128 or @llvm.x86.vcvtph2ps.256 to handle the
+        // conversion
+        internal_assert(srcTy.width == destTy.width) << "Source and destination widths must match\n";
+        if (srcTy.width <= 4) {
+            // Use @llvm.x86.vcvtph2ps.128. This returns <4 x float>
+            // FIXME: Codegen using other widths (e.g. 3) seems to give wrong
+            // results explicitly disable until this can be investigated.
+            internal_assert(srcTy.width == 4 || srcTy.width == 1) <<
+              "FIXME: Doing codegen with width less than four produces incorrect results\n";
+
+            // Codegen argument
+            llvm::Value* valueToCast = codegen(op->value);
+
+            // Note with width is 8 here because that is what is expected by
+            // the instrinsic. Only the first 4 lanes matter here.
+            // Slice to correct width
+            if (srcTy.width == 1) {
+                // Handle scalar value by inserting as the first element
+                // in a vector
+                internal_assert(!(valueToCast->getType()->isVectorTy()));
+                Constant* zeroHalfVectorSplat = ConstantFP::get(f16x8, 0.0);
+                valueToCast = builder->CreateInsertElement(/*Vec=*/zeroHalfVectorSplat,
+                                                           /*NewElt=*/valueToCast,
+                                                           /*Idx=*/ConstantInt::get(i32, 0)
+                                                          );
+            } else {
+                // Handle vectorized inputs by extending to correct width.
+                // Introduced undef lanes will be removed later.
+                valueToCast = slice_vector(valueToCast, /*start=*/0, /*size=*/8);
+            }
+
+            // Cast argument to <8 x i16> as expected by the intrinsic
+            valueToCast = builder->CreateBitCast(valueToCast, i16x8);
+
+            // Can't use call_intrin() because it asserts the number of elements
```

Comment:
@abadams It just occurred to me there is an additional issue here. On other platforms there may be a native instruction for conversion between ``float16`` and ``single``/``double`` but that might not be a vector instruction. An example is the ``FCVT`` instruction on AArch64. That means you need a way of picking the non vector hardware instruction over the software implementation and what you've proposed doesn't offer this IIUC.